From c98acf4f8506748fa4dfcf42e6582bedad9fa94c Mon Sep 17 00:00:00 2001
From: Jihong Min <hurryman2212@gmail.com>
Date: Tue, 21 Jul 2026 04:45:25 +0900
Subject: [PATCH] bonding: update XFRM offload across mode changes

Flush XFRM state while the bond mode or transmit hash policy
changes. Recompute the advertised XFRM features when entering or
leaving a supported LAG configuration.

Assisted-by: Codex:gpt-5.6
Signed-off-by: Jihong Min <hurryman2212@gmail.com>
---
 drivers/net/bonding/bond_options.c | 62 ++++++++++++++++++++++++++++--
 1 file changed, 59 insertions(+), 3 deletions(-)

diff --git a/drivers/net/bonding/bond_options.c b/drivers/net/bonding/bond_options.c
index adc216df4345..cb429b9368c0 100644
--- a/drivers/net/bonding/bond_options.c
+++ b/drivers/net/bonding/bond_options.c
@@ -16,6 +16,7 @@
 
 #include <net/bonding.h>
 #include <net/ndisc.h>
+#include <net/xfrm.h>
 
 static int bond_option_active_slave_set(struct bonding *bond,
 					const struct bond_opt_value *newval);
@@ -879,23 +880,45 @@ const struct bond_option *bond_opt_get(unsigned int option)
 
 static bool bond_set_xfrm_features(struct bonding *bond)
 {
+	netdev_features_t old_wanted;
+
 	if (!IS_ENABLED(CONFIG_XFRM_OFFLOAD))
 		return false;
 
-	if (BOND_MODE(bond) == BOND_MODE_ACTIVEBACKUP)
+	old_wanted = bond->dev->wanted_features;
+	if (BOND_MODE(bond) == BOND_MODE_ACTIVEBACKUP ||
+	    bond_mode_can_use_lag_xfrm(bond))
 		bond->dev->wanted_features |= BOND_XFRM_FEATURES;
 	else
 		bond->dev->wanted_features &= ~BOND_XFRM_FEATURES;
 
-	return true;
+	return (old_wanted ^ bond->dev->wanted_features) & BOND_XFRM_FEATURES;
 }
 
 static int bond_option_mode_set(struct bonding *bond,
 				const struct bond_opt_value *newval)
 {
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	bool old_lag_xfrm = bond_mode_can_use_lag_xfrm(bond);
+	bool new_lag_xfrm;
+	bool flush_xfrm;
+#endif
+
 	if (bond->xdp_prog && !bond_xdp_check(bond, newval->value))
 		return -EOPNOTSUPP;
 
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	new_lag_xfrm = (newval->value == BOND_MODE_8023AD ||
+			newval->value == BOND_MODE_XOR) &&
+			bond->params.xmit_policy == BOND_XMIT_POLICY_LAYER34;
+	flush_xfrm = old_lag_xfrm || new_lag_xfrm;
+	if (flush_xfrm) {
+		bond_ipsec_lag_begin_flush(bond);
+		if (bond->dev->reg_state == NETREG_REGISTERED)
+			xfrm_dev_state_flush(dev_net(bond->dev), bond->dev, true);
+	}
+#endif
+
 	if (!bond_mode_uses_arp(newval->value)) {
 		if (bond->params.arp_interval) {
 			netdev_dbg(bond->dev, "%s mode is incompatible with arp monitoring, start mii monitoring\n",
@@ -937,6 +960,11 @@ static int bond_option_mode_set(struct bonding *bond,
 
 	bond_xdp_set_features(bond->dev);
 
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	if (flush_xfrm)
+		bond_ipsec_lag_end_flush(bond);
+#endif
+
 	return 0;
 }
 
@@ -1574,11 +1602,39 @@ static int bond_option_fail_over_mac_set(struct bonding *bond,
 static int bond_option_xmit_hash_policy_set(struct bonding *bond,
 					    const struct bond_opt_value *newval)
 {
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	bool old_lag_xfrm = bond_mode_can_use_lag_xfrm(bond);
+	bool new_lag_xfrm;
+	bool flush_xfrm;
+#endif
+
 	if (bond->xdp_prog && !__bond_xdp_check(BOND_MODE(bond), newval->value))
 		return -EOPNOTSUPP;
 	netdev_dbg(bond->dev, "Setting xmit hash policy to %s (%llu)\n",
 		   newval->string, newval->value);
-	bond->params.xmit_policy = newval->value;
+
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	new_lag_xfrm = (BOND_MODE(bond) == BOND_MODE_8023AD ||
+			BOND_MODE(bond) == BOND_MODE_XOR) &&
+			newval->value == BOND_XMIT_POLICY_LAYER34;
+	flush_xfrm = old_lag_xfrm || new_lag_xfrm;
+	if (flush_xfrm) {
+		bond_ipsec_lag_begin_flush(bond);
+		if (bond->dev->reg_state == NETREG_REGISTERED)
+			xfrm_dev_state_flush(dev_net(bond->dev), bond->dev, true);
+	}
+#endif
+
+	WRITE_ONCE(bond->params.xmit_policy, newval->value);
+
+	if (bond->dev->reg_state == NETREG_REGISTERED &&
+	    bond_set_xfrm_features(bond))
+		netdev_update_features(bond->dev);
+
+#if IS_ENABLED(CONFIG_XFRM_OFFLOAD)
+	if (flush_xfrm)
+		bond_ipsec_lag_end_flush(bond);
+#endif
 
 	return 0;
 }
-- 
2.53.0
